### PR TITLE
Application authorization flow - Implement the alternative flow

### PR DESCRIPTION
**Pull request by normegil**
Hey,

I've implemented the alternative authentication flow used by desktop/mobile applications (Flow described here: https://docs.esi.evetech.net/docs/sso/native_sso_flow.html)

What it means is:
- The builder has a new option (enable_application_authentication)
- Client Secret is no longer mandatory IF the application flow is enabled. But at least one of them needs to be present.
- get_authorize_url return a struct. It's clearer that way, and I needed to return the pkce_verifier for the authenticate method.
-- I wanted to avoid managing extra state between each authentication request, especially since the pkce need to be regenerated for each new authentication flow.

I've tested it with my own app, and it works quite well.

I hope I didn't forgot to update documentation where it's required..

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -736,6 +783,20 @@ fn current_time_millis() -> Result<u128, EsiError> {
     Ok(SystemTime::now().duration_since(UNIX_EPOCH)?.as_millis())
 }
 
+/// AuthenticationInformations contains informations needed to complete the requested authentication flow.
+pub struct AuthenticationInformations {
+    /// URL to call/pass to users to initiate an authentication and get an auth code from ESI.
+    pub authorization_url: String,
+    /// If the default feature "random_state" is enabled, the returned state field string will be
+    /// random; otherwise it'll be "rfesi_unused". The ESI docs link to [this auth0 page] to
```
Comment (Celeo) [inline src/client.rs]: Missing URL for "this auth0 page"?

File: src/jwt_util.rs [hunk 2]
```diff
@@ -43,10 +43,14 @@ async fn get_rs256_key(client: &Client) -> EsiResult<String> {
 }
 
 /// Decode and validate the JWT token
-fn validate(token: &str, client_id: &str, decoding_key: &DecodingKey) -> Result<TokenClaims, EsiError> {
+fn validate(token: &str, client_id: Option<&str>, decoding_key: &DecodingKey) -> Result<TokenClaims, EsiError> {
```
Comment (Celeo) [inline src/jwt_util.rs]: I think this change might've snuck in from last time?

File: src/client.rs (hunk 1 above)
Comment (normegil) [reply src/client.rs]: Should be corrected. Probably a bad copy paste.

File: src/jwt_util.rs (hunk 2 above)
Comment (normegil) [reply src/jwt_util.rs]: Yeah. I think I'm doing something wrong with my pull request, but I'm not sure what. I've corrected the code again.

**Review by Celeo — APPROVED** [review]
Brilliant!

**Comment by Celeo** [comment]
Released in `0.29.0`.

**Comment by rafaga** [comment]
Sorry for asking here  but I'm not exactly sure how this affect my own desktop application? can someone put an actual example on how to use this new workflow in the correct way ?

**Comment by Celeo** [comment]
I haven't made a non-web application using ESI, but you should be able to use this to implement https://docs.esi.evetech.net/docs/sso/native_sso_flow.html and make such an application.

**Pull request merged**